### PR TITLE
RFE: add notification support

### PR DESCRIPTION
**Pull request by pcmoore**
This PR takes @tych0's original patch, adds the necessary Python support, sorts out some corner cases, and cleans up the API a bit.  @tych0 I left your sign-off on the main patch and the test patch as they inherited a large chunk of your code, please let me know if that's okay.

### CODE REVIEW

**Comment by tych0** [comment]
+1, sounds good to me

**Comment by drakenclimber** [comment]
I can help look through this tomorrow.  Thanks @pcmoore and @tych0

**Comment by pcmoore** [comment]
@drakenclimber thanks, although check the comments before you start - I'm currently arm wrestling with the old build environment in Travis.

**Comment by pcmoore** [comment]
Okay, it looks like the old header file problem is fixed now.

File: tests/51-live-user_notification.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
+#include <sys/types.h>
```
Comment (drakenclimber) [inline tests/51-live-user_notification.c]: nitpick - do you want to add your copyright info at the top?

File: include/seccomp.h.in [hunk 2]
```diff
@@ -673,6 +699,72 @@ int seccomp_rule_add_exact_array(scmp_filter_ctx ctx,
...
+int seccomp_notify_alloc(struct seccomp_notif **req,
+			 struct seccomp_notif_resp **resp);
+
+/**
+ * Free a pair of notification request/response structures.
+ * @param req the request location
+ * @param resp the response location
+ */
+void seccomp_notify_free(struct seccomp_notif *req,
+			 struct seccomp_notif_resp *resp);
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply include/seccomp.h.in]: Nice catch, fixed.  This will be in the next push.

File: src/system.c [hunk 3]
```diff
@@ -291,10 +321,17 @@ int sys_filter_load(const struct db_filter_col *col)
 			flgs |= SECCOMP_FILTER_FLAG_LOG;
 		if (col->attr.spec_allow)
 			flgs |= SECCOMP_FILTER_FLAG_SPEC_ALLOW;
+		if (_support_seccomp_user_notif > 0)
+			flgs |= SECCOMP_FILTER_FLAG_NEW_LISTENER;
 		rc = syscall(_nr_seccomp, SECCOMP_SET_MODE_FILTER, flgs, prgm);
 		if (rc > 0 && col->attr.tsync_enable)
 			/* always return -ESRCH if we fail to sync threads */
 			errno = ESRCH;
+		if (rc > 0 && _support_seccomp_user_notif > 0) {
```
[the remark this answers is not in the document]
Comment (tych0) [reply src/system.c]: The kernel doesn't allow setting both flags at the same time, so I think it should be ok.

[the remark this answers is not in the document]
Comment (pcmoore) [reply src/system.c]: Another good catch.  I just talked to @tych0 about this, and it looks like the kernel doesn't allow setting both TSYNC and NEW_LISTENER at the same time so we should be safe here.

I'm going to add another patch to this patchset to explicitly disallow this combination in libseccomp.

File: src/system.c [hunk 4]
```diff
@@ -303,5 +340,70 @@ int sys_filter_load(const struct db_filter_col *col)
...
+	if (req) {
+		*req = zmalloc(sizes.seccomp_notif);
+		if (!*req)
+			return -ENOMEM;
+	}
+
+	if (resp) {
+		*resp = zmalloc(sizes.seccomp_notif_resp);
+		if (!*resp) {
+			if (req)
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/system.c]: Yeah, I thought the same thing, but since there is only one instance I decided to keep it as-is.

File: src/system.c [hunk 5]
```diff
@@ -303,5 +340,70 @@ int sys_filter_load(const struct db_filter_col *col)
...
+	static struct seccomp_notif_sizes sizes = { 0, 0, 0 };
+
+	if (_support_seccomp_syscall <= 0)
+		return -EOPNOTSUPP;
+
+	if (sizes.seccomp_notif == 0 && sizes.seccomp_notif_resp == 0) {
+		rc = syscall(__NR_seccomp, SECCOMP_GET_NOTIF_SIZES, 0, &sizes);
+		if (rc < 0)
+			return -errno;
+	} else if (sizes.seccomp_notif == 0 || sizes.seccomp_notif_resp == 0)
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/system.c]: Good point, we should also check to make sure both of those values are non-zero.  Fixed in the next push.

File: src/python/seccomp.pyx [hunk 6]
```diff
@@ -343,6 +351,189 @@ cdef class Arg:
...
+                self._data.data.args[4],
+                self._data.data.args[5]]
+        return args
+
+cdef class NotificationResponse:
+    """ Python object representing a seccomp notification response.
+    """
+    cdef libseccomp.seccomp_notif_resp _data
+
+    def __cinit__(self, notify, val = 0, error = 0 , flags = 0):
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/python/seccomp.pyx]: Good catch, fixed.

File: src/python/seccomp.pyx [hunk 7]
```diff
@@ -711,6 +902,57 @@ cdef class SyscallFilter:
...
+        the NOTIFY action.
+        """
+        cdef libseccomp.seccomp_notif *req
+
+        fd = libseccomp.seccomp_notify_fd(self._ctx)
+        if fd < 0:
+            raise RuntimeError("Notifications not enabled/active")
+        rc = libseccomp.seccomp_notify_alloc(&req, NULL)
+        if rc < 0:
+            raise RuntimeError("Unable to allocate notification structures")
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/python/seccomp.pyx]: Sure, I'll switch it over to the "Library error (errno = X)" exception that we use elsewhere.

File: src/python/seccomp.pyx [hunk 8]
```diff
@@ -343,6 +351,189 @@ cdef class Arg:
...
+        self._data.data.nr = syscall
+        self._data.data.instruction_pointer = ip
+        self._data.data.args[0] = args[0]
+        self._data.data.args[1] = args[1]
+        self._data.data.args[2] = args[2]
+        self._data.data.args[3] = args[3]
+        self._data.data.args[4] = args[4]
+        self._data.data.args[5] = args[5]
+
+    def get_id(self):
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/python/seccomp.pyx]: I need to preface my comments here by saying I am *DEFINITELY* not Python expert; to be brutally honest, I can barely write Python.

That said, the property decorator seems like an awfully heavyweight solution to this problem.  The Notification object returns a standard Python list via the get_syscall_args() method so users of the Python bindings should be able to interact with the Notification object using a "pythonic" interface; this really only affects the binding implementation and isn't a general problem.

FWIW, the reason we have to do these assignments in the first place is that Cython isn't able to convert between a uint64_t array and a Python list; other array types do work (e.g. unsigned long), but since the args structure is a kernel struct I'm not sure I want to mess around to much with the types.

File: tests/51-live-user_notification.c [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		goto out;
+
+	rc = seccomp_notify_receive(fd, req);
+	if (rc)
+		goto out;
+	if (req->data.nr != SCMP_SYS(getpid)) {
+		rc = -EFAULT;
+		goto out;
+	}
+
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply tests/51-live-user_notification.c]: Good point, I'll change this for the next push.

File: tests/51-live-user_notification.c (hunk 1 above)
Comment (pcmoore) [reply tests/51-live-user_notification.c]: Good point, fixed.

File: src/python/seccomp.pyx (hunk 8 above)
[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/python/seccomp.pyx]: Sounds reasonable to me.  I haven't run the following, but I think it would provide a lighter-weight (if I am understanding correctly) `@property` solution:

```
   @property
   def id(self):
       return self._data.id

    # no need to have an id.setter

...

   @property
   def syscall_args(self):
       args = [self._data.data.args[0],
                self._data.data.args[1],
                self._data.data.args[2],
                self._data.data.args[3],
                self._data.data.args[4],
                self._data.data.args[5]]
        return args
```

[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/python/seccomp.pyx]: Then a user like `respond_notify()` can access it via `resp.id = response.id`

[the remark this answers is not in the document]
Comment (pcmoore) [reply src/python/seccomp.pyx]: Hmm, okay, I think I misunderstood both your original comment and Python properties.  Let me take a crack at this ...

[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/python/seccomp.pyx]: I've had good luck with `@property` in the past.  But I admit that I haven't used it with Cython, though I don't see why that would matter.

[the remark this answers is not in the document]
Comment (pcmoore) [reply src/python/seccomp.pyx]: Okay, check out the latest push, I believe it has the all property stuff done correctly.  I *think*.

File: src/db.c [hunk 10]
```diff
@@ -1150,30 +1154,39 @@ void db_col_release(struct db_filter_col *col)
...
 }
 
 /**
- * Validate a filter collection
- * @param col the seccomp filter collection
- *
- * This function validates a seccomp filter collection.  Returns zero if the
- * collection is valid, negative values on failure.
+ * Validate the seccomp action
+ * @param action the seccomp action
```
Comment (drakenclimber) [inline src/db.c]: nitpick - we should add @param for col as well

Comment (pcmoore) [reply src/db.c]: Another good catch, thanks.

Since this is the only remaining issue and it's a function header comment for an internal function, I'm simply going to go ahead and merge this after I fix it up, another review round seems a little silly at this point.  Thanks for all the review help.

**Comment by pcmoore** [comment]
Closing this out now, all the patches have been merged into the master branch.  Thanks @tych0 and @drakenclimber.
